api/s2n.h

```diff
@@ -2836,7 +2836,7 @@ S2N_API int s2n_psk_set_hmac(struct s2n_psk *psk, s2n_psk_hmac hmac);
  * @param conn A pointer to the s2n_connection object that contains the list of PSKs supported.
  * @param psk A pointer to the `s2n_psk` object to be appended to the list of PSKs on the s2n connection.
  */
-S2N_API int s2n_connection_append_psk(struct s2n_connection *conn, struct s2n_psk *psk);
+S2N_API int s2n_connection_append_psk(struct s2n_connection *conn, const struct s2n_psk *psk);
 
```

Comment:
My concern is that we're just marking methods that aren't currently affected by the blob / stuffer limitations as const. It's a fairly arbitrary subset, and if the implementation of one of these methods changes then we could end up having to [hackily ignore the const](https://github.com/aws/s2n-tls/blob/main/tls/s2n_send.c#L269), which in my opinion is worse than not having the const in the first place because the caller may make unsafe assumptions. I think we need to solve our actual limitations before we start trying to mark things const.

Reply:
> and if the implementation of one of these methods changes then we could end up having to [hackily ignore the const](https://github.com/aws/s2n-tls/blob/main/tls/s2n_send.c#L269),

I'm not sure I'm totally following. 
1. Actual implementation change so that we mutate data that was previously const
I don't think you're talking about this case, but just to be clear this would be a breaking change from our side, and we couldn't safely ignore the const designation.

2. C struct shenanigans make things difficult
The classic example here is the stuffer, which takes in an `s2n_blob` with a mutable reference to the underlying data, so we might end up casting to a mutable type to use the stuffer and then just manually type checking everything. 

I agree that it's tedious and dangerous to cast away the constness, but _we are already doing that_. Our Rust bindings currently have to cast away away the constness of types to call the underlying C methods.

https://github.com/aws/s2n-tls/blob/76af4d07883a3bf938f2d111919989542d187bf5/bindings/rust/s2n-tls/src/client_hello.rs#L106-L113

I would argue that these rust casts are more dangerous, because they are doing so with less context of the underlying implementation. By forcing the C code to handle these casts, we are pushing the unsafe code into a smaller context that can be manually reviewed more easily.

In summary I agree that this isn't a perfect solution. Calling across the C FFI will always be dangerous (C doesn't have transitive constness on pointers 😱 ), but I do think that this is a strict improvement on the safety of our bindings with very little cost.

Reply:
> I think we need to solve our actual limitations before we start trying to mark things const.

I guess constraining the discussion to these specific APIs, are you arguing that we should make the Rust bindings take mutable references every place that we have a mutable pointer?

Reply:
I'm only worried about your case #2, where the behavior of the methods should still be const but we can't actually handle const variables properly.

My argument is that we're currently already casting away the const-ness in Rust, but NOT in C. So this doesn't really help Rust if we just end up ignoring const anyway, and it's worse for C. The fact that these methods don't take const is a warning to C applications that we're not actually handling them as const.

I am not arguing that we should make the Rust bindings take mutable references where the underlying C api takes a mutable pointer. Unlike C, Rust actually enforces rules around mutability, so that would result in a completely unusable api. Our policy has been to constrain the casting to our Rust bindings. We make a lot of assumptions around safety in our Rust bindings-- just see my notes around Send/Sync safety.

Reply:
> The fact that these methods don't take const is a warning to C applications that we're not actually handling them as const.

I don't totally agree with this. By letting the Rust bindings call these APIs from an underlying const reference, we are making the promise that these C apis are "logically const", independent of the types that we are actually using. While we might not be currently using `const` types, we absolutely must be handling the data as "logically const", otherwise our Rust bindings are unsound/unsafe.

> We make a lot of assumptions around safety in our Rust bindings

I also wouldn't quite call these "assumptions". We manually check our C api's to ensure that they are truly logically const, and therefore that our Rust bindings are safe.

> So this doesn't really help Rust if we just end up ignoring const anyway

Right now we have to do manual validation that the mutable types are only used in ways that preserve the logical constness of APIs. We are only human, and might do this manual validation wrong. By declaring the apis as const, we can offload some of that analysis to the type system/compiler. This compiler doesn't get this wrong, so it helps the Rust bindings by making them safer.

Side Note: we cannot ignore that these apis are "logically const", even if the types themselves are mutable, but I think you're talking about `const` strictly in the type sense in the above comment.